File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;

        TreeNode curr = root;

        // ensure p < q
        if (q.val > p.val) {
            TreeNode tmp = p;
            p = q;
            q = tmp;
        }

        // we're guaranteed to return from this loop
        while (curr != null) {
            if (curr.val == p.val || curr.val == q.val) {
                return curr;
            }
            if (p.val < curr.val && q.val < curr.val) {
                curr = curr.left;
            } else if (p.val > curr.val && q.val > curr.val) {
                curr = curr.right;
            } else {
                return curr;
            }
        }

        return null;
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        if (!root) {
            return 0;
        }

        const index = [1];

        return this.inOrderTraversal(root, index, k);
    }

    inOrderTraversal(node, index, k) {
        if (!node) {
            return null;
        }

        const left = this.inOrderTraversal(node.left, index, k);

        if (left !== null) {
            return left;
        }

        if (index[0] === k) {
            return node.val;
        }
        // only increment AFTER we determine that we are not at the current index ("post processing")
        index[0]++;
        
        return this.inOrderTraversal(node.right, index, k);
    }
}
